libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -0,0 +1,528 @@
+flystel_prime_field_gadget<FieldT, generator>::flystel_prime_field_gadget(
+    protoboard<FieldT> &pb,
+    const linear_combination<FieldT> &x0,
+    const linear_combination<FieldT> &x1,
+    const pb_variable<FieldT> &y0,
+    const pb_variable<FieldT> &y1,
+    const std::string &annotation_prefix)
+    : gadget<FieldT>(pb, annotation_prefix)
+    , a0(pb_variable_allocate(pb, FMT(annotation_prefix, " a0")))
+    , a1(pb_variable_allocate(pb, FMT(annotation_prefix, " a1")))
+    , a2(pb_variable_allocate(pb, FMT(annotation_prefix, " a2")))
+    , input_x0(x0)
+    , input_x1(x1)
+    , output_y0(y0)
+    , output_y1(y1)
+    , Q_gamma(pb, x1, a0, FMT(annotation_prefix, " Q_gamma"))
+    , Q_delta(pb, x1 - a1, a2, FMT(annotation_prefix, " Q_delta"))
+    , E_root_five(pb, x0 - a0, a1, FMT(annotation_prefix, " E_root_five"))
+{
+    const FieldT input_x0_value =
+        input_x0.evaluate(this->pb.full_variable_assignment());
+    const FieldT input_x1_value =
+        input_x1.evaluate(this->pb.full_variable_assignment());
+    const FieldT a0_value = this->pb.val(a0);
+
+    printf("[%s:%d] %s() x0 = ", __FILE__, __LINE__, __FUNCTION__);
+    input_x0_value.print();
+    printf("[%s:%d] %s() x1 = ", __FILE__, __LINE__, __FUNCTION__);
+    input_x1_value.print();
+    printf("[%s:%d] %s() a0 = ", __FILE__, __LINE__, __FUNCTION__);
+    a0_value.print();
+}
+
+template<typename FieldT, size_t generator>
+void flystel_prime_field_gadget<FieldT, generator>::generate_r1cs_constraints()
+{
+    Q_gamma.generate_r1cs_constraints();
+    Q_delta.generate_r1cs_constraints();
+    E_root_five.generate_r1cs_constraints();
+}
+
+template<typename FieldT, size_t generator>
+void flystel_prime_field_gadget<FieldT, generator>::generate_r1cs_witness()
+{
+    Q_gamma.generate_r1cs_witness();
+    E_root_five.generate_r1cs_witness();
+    Q_delta.generate_r1cs_witness();
+
+    const FieldT input_x0_value =
+        input_x0.evaluate(this->pb.full_variable_assignment());
+    const FieldT input_x1_value =
+        input_x1.evaluate(this->pb.full_variable_assignment());
+
+    this->pb.lc_val(output_y0) =
+        input_x0_value - this->pb.val(a0) + this->pb.val(a2);
+    this->pb.lc_val(output_y1) = input_x1_value - this->pb.val(a1);
+
+    printf("[%s:%d] x0  ", __FILE__, __LINE__);
+    input_x0_value.print();
+    printf("[%s:%d] a0  ", __FILE__, __LINE__);
+    this->pb.val(a0).print();
+    printf("[%s:%d] a2  ", __FILE__, __LINE__);
+    this->pb.val(a2).print();
+    printf("[%s:%d] y0  ", __FILE__, __LINE__);
+    this->pb.lc_val(output_y0).print();
+    printf("[%s:%d] y1  ", __FILE__, __LINE__);
+    this->pb.lc_val(output_y1).print();
+}
+
+template<typename FieldT, size_t NumStateColumns_L>
+std::array<std::array<FieldT, NumStateColumns_L>, NumStateColumns_L>
+anemoi_permutation_mds(const FieldT g)
+{
+    std::array<std::array<FieldT, NumStateColumns_L>, NumStateColumns_L> M;
+    const FieldT g2 = g * g;
+    if (NumStateColumns_L == 2) {
```

Comment:
I think we can just provide implementations of this function for the allowed values of `NumStateColumns_L`.  These values are known at compile time, so we should not have to wait until runtime.  Perhaps a `static_assert` can be used?

Reply:
#77

Reply:
TBD in separate branch